pkg/types/query/query.go

```diff
@@ -0,0 +1,208 @@
+	BoolOperator
+}
+
+// Comparator is used for filtering through specific key values.
+// e.g. of filtering for key that belongs to a token transfer by values: Comparator("transferValue", [{"150",LTE}, {"300",GTE}])
+func Comparator(name string, valueComparators ...primitives.ValueComparator) Expression {
+	return Expression{Primitive: &primitives.Comparator{Name: name, ValueComparators: valueComparators}}
+}
+
+func Block(block uint64, operator primitives.ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &primitives.Block{Block: block, Operator: operator},
+	}
+}
+
+func Confirmation(confLevel primitives.ConfirmationLevel) Expression {
+	return Expression{
+		Primitive: &primitives.Confirmations{ConfirmationLevel: confLevel},
+	}
+}
+
+func Timestamp(timestamp uint64, operator primitives.ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &primitives.Timestamp{Timestamp: timestamp, Operator: operator},
+	}
+}
+
+func TxHash(txHash string) Expression {
+	return Expression{
+		Primitive: &primitives.TxHash{TxHash: txHash},
+	}
+}
+
+func And(expressions ...Expression) Expression {
+	return Expression{
+		BoolExpression: BoolExpression{Expressions: expressions, BoolOperator: AND},
+	}
+}
+
+func Or(expressions ...Expression) Expression {
+	return Expression{
+		BoolExpression: BoolExpression{Expressions: expressions, BoolOperator: OR},
+	}
+}
+
+// Where is a helper function for building KeyFilter, eg. usage:
+//
+//	queryFilter, err := Where(
+//			"key",
+//			TxHash("0xHash"),
+//			Block(startBlock, Gte),
+//			Block(endBlock, Lte),
+//			Or(
+//				And(
+//					Timestamp(someTs1, Gte),
+//					Timestamp(otherTs1, Lte)),
+//				And(
+//					Timestamp(someTs2, Gte),
+//					Timestamp(otherTs2, Lte)),
+//			),
+//		)
+//	Equals:`txHash = '0xHash' AND
+//			block > startBlock AND
+//			block < endBlock   AND
+//			((timestamp > someTs1 And timestamp < otherTs1)
+//				OR
+//			(timestamp > someTs2 And timestamp < otherTs2))`
+func Where(key string, expressions ...Expression) (KeyFilter, error) {
+	for _, expr := range expressions {
+		if !expr.IsPrimitive() {
+			if len(expr.BoolExpression.Expressions) < 2 {
+				return KeyFilter{}, fmt.Errorf("all boolean expressions should have at least 2 expressions")
+			}
+		}
+	}
+	return KeyFilter{Key: key, Expressions: expressions}, nil
+}
+
+type SortDirection int
+
+const (
+	Asc SortDirection = iota
+	Desc
+)
+
+type SortBy interface {
+	GetDirection() SortDirection
+}
+
+type CursorDirection int32
+
+const (
+	CursorPrevious CursorDirection = iota + 1
+	CursorFollowing
+)
+
+type Limit struct {
+	Cursor          string
+	CursorDirection CursorDirection
+	Count           uint64
+}
+
+func CursorLimit(cursor string, cursorDirection CursorDirection, count uint64) Limit {
+	return Limit{
+		Cursor:          cursor,
+		CursorDirection: cursorDirection,
+		Count:           count,
+	}
+}
+
+func CountLimit(count uint64) Limit {
+	return Limit{Count: count}
```

Comment:
Without the pointer fields, do these constructors still add value? Or would directly initializing the structs at the calls sites be more clear?
```diff
-limit := query.CursorLimit("foo", query.CursorPrevious, 10)
+limit := query.Limit{Cursor: "foor", CursorDirection: query.CursorPrevious, Count: 10)
```
```diff
-limit := query.CountLimit(10)
+limit := query.Limit{Count: 10)
```

Reply:
Almost everything else is initialised via constructros, I am leaning towards leaving it like this for consitency

Reply:
Usually constructors are the exception. They also will become inflexible once we cut v1.0.0 of this repo FWIW :shrug: